Internal #2017: DECIMAL Downcast Rounding

hawkfish opened the pull request:
When we downcast DECIMAL values (reduce the precision) we need to round instead of truncating.

fixes: duckdblabs/duckdb-internal#2017

hawkfish committed:
Internal #2017: DECIMAL Downcast Rounding

When we downcast DECIMAl values (reduce the precision) we need to round
instead of truncating.

fixes: duckdblabs/duckdb-internal#2017

Mytherin commented on test/sql/types/decimal/test_decimal.test:
```diff
@@ -114,6 +114,17 @@ select '0.00003'::DECIMAL(38,10)::VARCHAR
 
+# Downcasting should round
+query I
+select CAST(33.846 AS DECIMAL(5,0)) d0;
```
Can we add a test when the rounding results in an overflow for the decimal type, e.g. in Postgres:

```sql
select cast('9.99' as decimal(1,0));
--- ERROR:  numeric field overflow
--- DETAIL:  A field with precision 1, scale 0 must round to an absolute value less than 10^1.

```

hawkfish replied:
Turns out this doesn't hit the new code.

Mytherin replied:
Have looked at this again - the reason it doesn't hit the new code is because there are two decimal operators - `DecimalScaleDownOperator` and `DecimalScaleDownCheckOperator`. Currently the implementation only exists for `DecimalScaleDownOperator` - which means the rounding will only happen when `source_width >= target_width`, i.e. when going down in precision no rounding happens. Perhaps this can be extended to work in both cases?

hawkfish replied:
It's in the string casting code actually. That does wasn't checking whether rounding overflowed.

Mytherin reviewed (commented):
Thanks for the PR! Looks good - one request for another test:

hawkfish committed:
Merge branch 'main' into decimal-rounding

hawkfish committed:
Merge branch 'main' into decimal-rounding

hawkfish committed:
Merge branch 'main' into decimal-rounding

hawkfish committed:
Internal #2017: VARCHAR DECIMAL Limits

Add a limit check in casting VARCHAR to DECIMAL to make sure
the rounding didn't overflow.

fixes: duckdblabs/duckdb-internal#2017

Commit:
Merge branch 'feature' into decimal-rounding

Commit:
Merge branch 'feature' into decimal-rounding

Commit:
Merge branch 'feature' into decimal-rounding

Mytherin commented:
Thanks!
